src/runtime-restriction.cpp

```diff
@@ -0,0 +1,105 @@
+#include "runtime-restriction.hpp"
+#include "elf-utils.hpp"
+#include "log.h"
+
+#include <elf.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unordered_map>
+
+#define PAGE_START(addr) (PAGE_MASK & addr)

```

Comment:
This does not align_down the pointer, but rather will truncate all _upper_ bits. For example, assuming a page size of 4KB:
`PAGE_MASK` is: `0xfff`, if you have an address that is: `0x7fb00da0` you will get: `0xda0` after masking. If you want to align down instead, use the bitwise not of the mask when anding it: `~PAGE_MASK & addr` which would have gotten you: `0x7fb00000` instead (which is page aligned)

Reply:
```
#define PAGE_SIZE 4096
#define PAGE_MASK (~(PAGE_SIZE - 1))
```
this aleady has the `~` infront so is it still needed?